hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -11,145 +14,127 @@
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE PartialTypeSignatures #-}
-module Torch.Data.Pipeline ( takeBatch
-                           , readBatches
-                           , readBatchesConcurrently
-                           , makeFold'
-                           , makeFoldWithTransform'
-                           , makeConcurrentFold'
-                           , makeFold
-                           , makeFoldWithTransform
-                           , makeConcurrentFold
-                           , foldFromProducer
-                           , L.FoldM(..)
-                           , Dataset(..)
-                           , DatasetMock(..)
-                           , ConcurrentDataset(..)
+module Torch.Data.Pipeline ( Dataset(..)
+                           , MapStyleOptions(..)
+                           , Sample(..)
+                           , makeListT
+                           , mapStyleOpts
+                           , runContT
                            ) where
 
+import           Control.Applicative (Alternative, (<|>))
 import           Control.Concurrent.Async.Lifted
 import qualified Control.Foldl as L
 import           Control.Monad
-import Control.Applicative (Alternative, (<|>))
-import           Data.Maybe (isJust)
+import           Data.Maybe (fromJust, isJust)
 import           Pipes
 import           Pipes.Concurrent
 import qualified Pipes.Prelude as P
 
 import           Control.Monad.Trans.Control (MonadBaseControl(..))
+import           Control.Monad.Cont (runContT, ContT)
+import           Torch.Data.Internal
+import qualified Torch as Torch
+import Lens.Family (view)
+import Pipes.Group
+import Data.Vector (Vector)
+import qualified Data.Map.Strict as M
+import Data.Set
+import Control.Concurrent.STM
+import qualified Data.Vector as V
+import Control.Monad.Base (MonadBase)
+import System.Random
+import qualified Data.IntMap as I
+import Data.IntMap (IntMap)
+import Control.Arrow (Arrow(first))
+
+data MapStyleOptions = MapStyleOptions { bufferSize :: Int
+                                       , numWorkers :: Int
+                                       , shuffle :: Sample
+                                       }
+
+mapStyleOpts numWorkers = MapStyleOptions { bufferSize = numWorkers
+                                          , numWorkers = numWorkers
+                                          , shuffle = Sequential
+                                          }
+
+data Sample = Sequential | Shuffle StdGen
```

Comment:
```suggestion
data Sample = Sequential | RandomGen g => Shuffle g
```
?
I'm actually not sure if this valid syntax for ADTs.

Reply:
I changed it to a GADT with this constraint